geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java

```diff
@@ -1637,8 +1640,13 @@ private void readMessages() {
               }
             }
             isHandShakeReader = true;
-            // Once we have read the handshake the reader can go away
-            break;
+            notifyHandshakeWaiter(false);
+            // Once we have read the handshake for unshared connections, the reader can skip
+            // processing messages
+            if (!sharedResource) {
```

Comment:
good change here.  It shouldn't be needed since a sharedResource will only invoke this method if isReaderThread==true but having an explicit test is a good idea.